File: python/탐욕법(그리디)/큰수만들기.py

""" Level 2. 큰 수 만들기 """

def solution(number, k):
    answer = []
    
    for num in range(0, len(number)):

        temp = number[num]

        if not answer:
            answer.append(temp)
            continue

        while answer and answer[-1] < temp and k > 0:
            answer.pop()
            k -= 1

        answer.append(temp)
    
    return ''.join(answer if k == 0 else answer[:-k])

print(solution("1924", 2))              # "94"
print(solution("1231234",3))            # "3234"
print(solution("4177252841", 4))        # "775841"

print(solution("1924", 3))              # "9"
print(solution("1231234",6))            # "4"
print(solution("4177252841", 9))        # "8"

File: Java/src/그리디/큰수만들기.java
package 그리디;

import java.util.*;

public class 큰수만들기 {
    
    public static void main(String[] args) {

        String n1 = "1924";
        int k1 = 2;
        System.out.println(solution(n1, k1));

        String n2 = "1231234";
        int k2 = 3;
        System.out.println(solution(n2, k2));

        String n3 = "4177252841";
        int k3 = 4;
        System.out.println(solution(n3, k3));
    }

    public static String solution(String number, int k) {

        Stack<Character> stack = new Stack<>();

        for (char temp : number.toCharArray()) {

            while (!stack.isEmpty() && stack.peek() < temp && k > 0) {
                stack.pop();
                k--;
            }

            stack.push(temp);
        }

        while (k > 0) {
            stack.pop();
            k--;
        }
        
        String answer = "";
        for (int i = 0; i < stack.size(); i++) {
            answer += stack.get(i);
        }

        return answer;
    }
}
